src/zmac.y

```diff
@@ -2318,50 +2338,34 @@ statement:
 	}
 |
 	IF_TK expression '\n' {
+		do_if($2);
+	}
+|
+	IFE_TK expression '\n' {
 		expr_number_check($2);
-		if (ifptr >= ifstmax)
-			error("Too many ifs");
-		else
-			*++ifptr = !($2->e_value);
-
-		saveopt = fopt;
-		fopt = 1;
-		list($2->e_value);
-		fopt = saveopt;
+		do_if_value(!$2->e_value);
 		expr_free($2);
 	}
+|
+	IF1_TK '\n' {
+		do_if_value(npass == 1);
+	}
+|
+	IF2_TK '\n' {
+		do_if_value(npass == 2);
+	}
 |
 	IF_CP_TK expression ',' expression '\n' {
 		// Unpleasant duplication of IF_TK work.
-		struct expr *compare = expr_mk($2, $1->i_value , $4);
-		expr_number_check(compare);
-		if (ifptr >= ifstmax)
-			error("Too many ifs");
-		else
-			*++ifptr = !(compare->e_value);
-
-		saveopt = fopt;
-		fopt = 1;
-		list(compare->e_value);
-		fopt = saveopt;
-		expr_free(compare);
+		do_if(expr_mk($2, $1->i_value , $4));
```

Comment:
You'll need to save the result of expr_mk() here so it can be freed with expr_free() after the do_if() call.

Reply:
There's an `expr_free()` inside the `do_if()` function at src/zmac.y:2139

Reply:
Oh, yes, you're right.  Sorry I missed that.